streams/src/main/java/org/apache/kafka/streams/kstream/Transformer.java

```diff
@@ -44,17 +46,18 @@
      *
      * @param key the key for the message
      * @param value the value for the message
-     * @return new value
+     * @return new value; if null no key-value pair will be forwarded to down stream
      */
-    R transform(K key, V value);
+    KeyValue<RK, RV> transform(K key, V value);
 
     /**
-     * Perform any periodic operations, if this processor {@link ProcessorContext#schedule(long) schedule itself} with the context
+     * Perform any periodic operations and possibly generate a new key and value, if this processor {@link ProcessorContext#schedule(long) schedule itself} with the context
      * during {@link #init(ProcessorContext) initialization}.
      *
      * @param timestamp the stream time when this method is being called
+     * @return new value; if null no key-value pair will be forwarded to down stream
      */
-    void punctuate(long timestamp);
+    KeyValue<RK, RV> punctuate(long timestamp);
```

Comment:
why don’t we return `Iterable<RK, RV>`? aren’t we limiting the punctuate to only return values for one key by doing this?

e.g., if i wanted `transform()` to do aggregation and only send values downstream on punctuate i'd have to do some pretty ugly flatmapping after this transform